qutebrowser/misc/overlap.py

```diff
@@ -0,0 +1,136 @@
+# vim: ft=python fileencoding=utf-8 sts=4 sw=4 et:
+
+# Copyright 2014-2021 Florian Bruhin (The Compiler) <mail@qutebrowser.org>
+#
+# This file is part of qutebrowser.
+#
+# qutebrowser is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# qutebrowser is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <https://www.gnu.org/licenses/>.
+
+"""Work with overlapping widgets."""
+
+import collections
+import math
+from typing import Dict, Iterable, List, Sequence, Set, Tuple
+
+from PyQt5.QtCore import QPoint, QRect
+from PyQt5.QtWidgets import QWidget
+
+
+def _overlap(r1: QRect, r2: QRect) -> bool:
+    hor_overlap = max(0,
+        (min(r1.right(), r2.right()) - max(r1.left(), r2.left()))
+    )
+    vert_overlap = max(0,
+        (min(r1.bottom(), r2.bottom()) - max(r1.top(), r2.top()))
+    )
+    return hor_overlap > 0 and vert_overlap > 0
+
+
+def _corners(widget: QWidget) -> List[QPoint]:
+    return [
+        widget.geometry().topLeft(), widget.geometry().topRight(),
```

Comment:
Here I'd probably use a variable for `widget.geometry()` to avoid having to call it again (and also because it's a bit unwieldy)